From 442e43df27c136192d4063da10aad79fd860ba67 Mon Sep 17 00:00:00 2001
From: Alberto Ponces <ponces26@gmail.com>
Date: Tue, 11 Aug 2026 08:46:23 +0000
Subject: [PATCH 2/7] Add option to change eSCO Transport Unit Size

Fixes Bluetooth calls on some Samsung devices if set to 16.

It is still unknown if other Treble devices need other values so it is preferred to leave this able to be configured with another integer.

This applies to mSBC T2, T1 and CVSD codecs

[ported to lineage-23.2/A16: esco_parameters.cc structure changed (LC3/SWB
paths); override inserted before the offload return block as in the original;
added libcutils to libbtdevice static_libs since property_get_int32 is a real
libcutils symbol in A16]

Change-Id: I3e5897c2ce983042b9a4bef9def6744ba4253bcb
---
 system/device/Android.bp             | 1 +
 system/device/src/esco_parameters.cc | 9 ++++++++-
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/system/device/Android.bp b/system/device/Android.bp
index de6d645c92..a915faccc1 100644
--- a/system/device/Android.bp
+++ b/system/device/Android.bp
@@ -31,6 +31,7 @@ cc_library_static {
     header_libs: ["libbluetooth_headers"],
     static_libs: [
         "bluetooth_flags_c_lib",
+        "libcutils",
         "libaconfig_storage_read_api_cc",
         "libbluetooth_hci_pdl",
         "libbluetooth_log",
diff --git a/system/device/src/esco_parameters.cc b/system/device/src/esco_parameters.cc
index ac60538bda..7d8c483590 100644
--- a/system/device/src/esco_parameters.cc
+++ b/system/device/src/esco_parameters.cc
@@ -24,9 +24,11 @@
 #include "hci/controller.h"
 #include "main/shim/entry.h"
 
+#include <cutils/properties.h>
+
 using namespace bluetooth;
 
-static const enh_esco_params_t default_esco_parameters[ESCO_NUM_CODECS] = {
+static enh_esco_params_t default_esco_parameters[ESCO_NUM_CODECS] = {
         // CVSD D1
         {
                 .transmit_bandwidth = TXRX_64KBITS_RATE,
@@ -382,6 +384,11 @@ enh_esco_params_t esco_parameters_for_codec(esco_codec_t codec, bool offload) {
     }
   }
 
+  int escoTransportUnitSize = property_get_int32("persist.sys.bt.esco_transport_unit_size", 0);
+  if (escoTransportUnitSize) {
+    default_esco_parameters[codec].input_transport_unit_size = escoTransportUnitSize;
+    default_esco_parameters[codec].output_transport_unit_size = escoTransportUnitSize;
+  }
   if (offload) {
     if (codec == ESCO_CODEC_SWB_Q0 || codec == ESCO_CODEC_SWB_Q1 || codec == ESCO_CODEC_SWB_Q2 ||
         codec == ESCO_CODEC_SWB_Q3) {
